orttraining/orttraining/core/session/training_session.cc

```diff
@@ -584,33 +608,73 @@ Status TrainingSession::AddTensorboard(const std::string& summary_name,
 
 Status TrainingSession::InsertPipelineOps(
     const std::unordered_set<std::string>& initializer_names_to_preserve,
-    std::string& forward_waited_event_name,
-    std::string& forward_recorded_event_name,
-    std::string& backward_waited_event_name,
-    std::string& backward_recorded_event_name,
-    std::string& forward_wait_output_name,
-    std::string& forward_record_output_name,
-    std::string& backward_wait_output_name,
-    std::string& backward_record_output_name,
-    std::string& forward_waited_event_after_recv_name,
-    std::string& forward_recorded_event_before_send_name,
-    std::string& backward_waited_event_after_recv_name,
-    std::string& backward_recorded_event_before_send_name) {
+    const std::string& loss_name,
+    // Forward Recv
+    std::string& forward_recv_waited_event_name,
+    std::string& forward_recv_wait_output_name,
```

Comment:
nit: might be better to use struct for function param. The params all have the same type and could easily mismatch in assignment without compiler help checking. With struct, it's easier using the named fields.

Reply:
Introduced a new class
```
struct PipelineTensorNames;
```
for containing all names.


---
In reply to: [476894314](https://github.com/microsoft/onnxruntime/pull/4916#discussion_r476894314) [](ancestors = 476894314)